Publications
============

Citation
--------

When citing Pyrodigal, please additionally cite the original Prodigal paper,
which developed the actual method. You can for instance cite both papers as:

    Pyrodigal (Larralde, 2022), a Python library binding to Prodigal (Hyatt et al., 2010).

Use the following for the bibliography:

- Hyatt, D., Chen, GL., LoCascio, P.F. et al. Prodigal: prokaryotic gene recognition and translation initiation site identification. BMC Bioinformatics 11, 119 (2010). :doi:`10.1186/1471-2105-11-119`.
- Larralde, M., (2022). Pyrodigal: Python bindings and interface to Prodigal, an efficient method for gene prediction in prokaryotes. Journal of Open Source Software, 7(72), 4296. :doi:`10.21105/joss.04296`.

Or just grab the `BibTex file <../_static/citation.bib>`_ with these two papers.


JOSS
----

Pyrodigal has been published in the `Journal of Open-Source Science <https://joss.theoj.org>`_.
JOSS is a community-run journal and relies upon volunteer effort. If you'd like
to support JOSS please consider doing either one (or both) of the the following:

- Volunteering to review for JOSS sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html.
- Making a small donation to support running costs here: https://numfocus.org/donate-to-joss.


Users
-----

Pyrodigal is being used in the following research works:

- Carroll, L. M., Larralde, M., Fleck, J. S., Ponnudurai, R., Milanese, A., Cappio, E., Zeller, G.  Accurate de novo identification of biosynthetic gene clusters with GECCO. bioRxiv 2021.05.03.442509; :doi:`10.1101/2021.05.03.442509`.
- Santos-Júnior C.D., Pan, S., Zhao, X., Coelho, L.P.  Macrel: antimicrobial peptide screening in genomes and metagenomes. PeerJ 8:e10555. :doi:`10.7717/peerj.10555`.
- Sirén K., Millard, A., Petersen, B., Gilbert, M. T. P., Clokie, M. RJ., Sicheritz-Pontén, T. Rapid discovery of novel prophages using biological feature engineering and machine learning. bioRxiv 2020.08.09.243022; :doi:`10.1101/2020.08.09.243022`.
- Turkington, C. J. R., Nezam Abadi, N., Edwards, R. A., Grasis, J. A.  hafeZ: Active prophage identification through read mapping. bioRxiv 2021.07.21.453177; :doi:`10.1101/2021.07.21.453177`.
- Reynolds, R., Hyun, S., Tully, B., Bien, J., Levine, N. M. Identification of microbial metabolic functional guilds from large genomic datasets. Front Microbiol. 2023 Jun 30;14:1197329. :doi:`10.3389/fmicb.2023.1197329`.
- Bouras, G., Grigson, S., Papudeshi, B., Mallawaarachchi, V., Roach, M.  Dnaapler: A tool to reorient circular microbial genomes. Journal of Open Source Software, 9(93), 5968. :doi:`10.21105/joss.05968`.
- Mainguy J, Hoede C. Binette: a fast and accurate bin refinement tool to construct high quality Metagenome Assembled Genomes. bioRxiv 2024.04.20.585171; :doi:`10.1101/2024.04.20.585171`.
- Ridley, R. S. Jr, Conrad, R. E., Lindner, B. G., Woo, S., Konstantinidis, K. T.  Potential routes of plastics biotransformation involving novel plastizymes revealed by global multi-omic analysis of plastic associated microbes. Sci Rep 14, 8798 (2024). :doi:`10.1038/s41598-024-59279-x`.
- Figueroa, J. L. III, Dhungel, E., Bellanger, M., Brouwer, C. R., White, R. A. III. MetaCerberus: distributed highly parallelized HMM-based processing for robust functional annotation across the tree of life. Bioinformatics, Volume 40, Issue 3, March 2024, btae119. :doi:`10.1093/bioinformatics/btae119`.
- Tesson, F., Planel, R., Egorov, A., Georjon, H., Vaysset, H., Brancotte, B., Néron, B. Mordret, E., Atkinson, G., Bernheim, A., Cury, J. A Comprehensive Resource for Exploring Antiphage Defense: DefenseFinder Webservice, Wiki and Databases. bioRxiv 2024.01.25.577194; :doi:`10.1101/2024.01.25.577194`.
- Cook, R., Telatin, A., Bouras, G., Camargo, A. P., Larralde, M., Edwards, R. A., Adriaenssens, E. M. Driving through stop signs: predicting stop codon reassignment improves functional annotation of bacteriophages, ISME Communications, Volume 4, Issue 1, January 2024, ycae079, :doi:`10.1093/ismeco/ycae079`.
- Bouras, G., Grigson, R., Papudeshi, B., Mallawaarachchi, V., Roach, M. J. Dnaapler: A tool to reorient circular microbial genomes. Journal of Open Source Software, 9(93), 5968. :doi:`10.21105/joss.05968`.
- Salamzade, R., Tran, P., Martin, C., Manson, A. L., Gilmore, M. S., Earl, A. M., Anantharaman, K., Kalan, L. R. zol & fai: large-scale targeted detection and evolutionary investigation of gene clusters. bioRxiv 2023.06.07.544063; :doi:`10.1101/2023.06.07.544063`.
- Duan, Y., Santos-Junior, C. D., Schmidt, T.S., Fullam, A., de Almeida, B. L. S., Zhu, C., Kuhn, M., Zhao, X.M., Bork, P., Coelho, L. P. A catalogue of small proteins from the global microbiome. bioRxiv 2023.12.27.573469; :doi:`10.1101/2023.12.27.573469`.
- Dimonaco, N. J., Clare, A., Kenobi, K., Aubrey, W., Creevey, C. J. StORF-Reporter: finding genes between genes, Nucleic Acids Research, Volume 51, Issue 21, 27 November 2023, Pages 11504–11517. :doi:`10.1093/nar/gkad814`.
- Reynolds, R., Hyun, S., Tully, B., Bien, J., Levine, N. M. Identification of microbial metabolic functional guilds from large genomic datasets. Front. Microbiol. 14:1197329. :doi:`10.3389/fmicb.2023.1197329`.
- Becker, J. W., Pollak, S., Berta-Thompson, J. W., Becker, K. W., Braakman, R., Dooley, K. D., Hackl, T., Coe, A., Arellano, A., LeGault, K. N., Berube, P. M., Biller, S. J., Cubillos-Ruiz, A., Van Mooy, B. A. S., Chisholm, S. W. Novel isolates expand the physiological diversity of Prochlorococcus and illuminate its macroevolution. bioRxiv 2023.12.03.569780; :doi:`10.1101/2023.12.03.569780`.
- Espinoza, J. L., Phillips, A., Prentice, M. B., Tan, G. S., Kamath, P. L., Lloyd, K. G., Dupont, C. L. Unveiling the microbial realm with VEBA 2.0: a modular bioinformatics suite for end-to-end genome-resolved prokaryotic, (micro)eukaryotic and viral multi-omics from either short- or long-read sequencing, Nucleic Acids Research, 2024;, gkae528, :doi:`10.1093/nar/gkae528`.
- Hahnfeld, J. M., Schwengers, O., Jelonek, L., Diedrich, S., Cemič, F., Goesmann, A. sORFdb – A database for sORFs, small proteins, and small protein families in bacteria. bioRxiv 2024.06.19.599710; :doi:`10.1101/2024.06.19.599710`.
- Schmitz, M., Dimonaco, N. J., Clavel, T., Hitch, T. C. A. Lineage-specific microbial protein prediction enables large-scale exploration of protein ecology within the human gut. bioRxiv 2024.05.29.596415; :doi:`10.1101/2024.05.29.596415`.
- Johnson, S. R., Weigele, P., Fomenkov, A., Ge, A., Vincze, A., Roberts, R. J., Sun, Z. Domainator, a flexible software suite for domain-based annotation and neighborhood analysis, identifies proteins involved in antiviral systems. bioRxiv 2024.04.23.590562; :doi:`10.1101/2024.04.23.590562`.
- Bouras, G., Houtak, G., Wick, R. R., Mallawaarachchi, V., Roach, M. J., Papudeshi, B., Judd, L. M., Sheppard, A. E., Edwards, R. A., Vreugde, S. Hybracter: enabling scalable, automated, complete and accurate bacterial genome assemblies. Microbial Genomics, Volume 10, Issue 5. :doi:`10.1099/mgen.0.001244`.

Feel free to get in touch to add yours to the list!
